### PR TITLE
Support `auth`

### PR DESCRIPTION
**Pull request by reynir**
A (nice-ish) parse error is raised if the hash is not one of those in mirage-crypto. Furthermore, `is_valid_config` checks that only `auth sha1` may be present for now.

### CODE REVIEW

File: src/config.ml [hunk 1]
```diff
@@ -240,9 +241,13 @@ module Conf_map = struct
...
-        if mem Cipher t && get Cipher t <> "AES-256-CBC" then
-          Error "currently only supported Cipher is 'AES-256-CBC'"
-        else Ok () )
+        (if mem Cipher t && get Cipher t <> "AES-256-CBC" then
+           Error "currently only supported Cipher is 'AES-256-CBC'"
+         else Ok ())
+        >>= fun () ->
+        match find Auth t with
+        | None | Some `SHA1 -> Ok ()
+        | Some _h -> Error "currently only supported auth is 'SHA1'" )
```
Comment (hannesm) [inline src/config.ml:250]: to me it looks like you've adjusted all the parts where `SHA1` was hardcoded... and my current guess it, any other (supported) hmac just works. did you try that?

File: src/config.ml [hunk 2]
```diff
@@ -379,7 +384,18 @@ module Conf_map = struct
...
     in
+    let pp_digest_algorithm ppf h =
+      Fmt.string ppf
+        (match h with
+        | `MD5 -> "MD5"
+        | `SHA1 -> "SHA1"
+        | `SHA224 -> "SHA224"
+        | `SHA256 -> "SHA256"
+        | `SHA384 -> "SHA384"
+        | `SHA512 -> "SHA512")
```
Comment (hannesm) [inline src/config.ml:395]: if we move this somwhere, we may reuse it in `State.pp`

File: src/engine.ml [hunk 3]
```diff
@@ -101,20 +105,28 @@ let hmacs config =
   match Config.find Tls_auth config with
   | None -> Error (`Msg "no tls auth payload in config")
   | Some (direction, _, hmac1, _, hmac2) ->
+      let hmac_len =
+        Option.value ~default:`SHA1 (Config.find Auth config)
+        |> Mirage_crypto.Hash.digest_size
+      in
```
Comment (hannesm) [inline src/engine.ml:111]: I'd propose to move ``Auth `SHA1`` into the default configuration, and avoid the `Option.value ~default` (that appears thrice in this PR). WDYT?

File: src/state.ml [hunk 4]
```diff
@@ -192,8 +193,9 @@ type session = {
 }
 
 let init_session ~my_session_id ?(their_session_id = 0L) ?(compress = false)
-    ?(protocol = `Tcp) ~my_hmac ~their_hmac () =
+    ?(protocol = `Tcp) ?(hmac_algorithm = `SHA1) ~my_hmac ~their_hmac () =
```
Comment (hannesm) [inline src/state.ml:196]: I wonder, is there negotiation of the `auth`? I.e. if a client has one in its configuration, can the server deselect it and choose a different one (i.e. for PUSH_NOTIFY)?

If the answer is no, why should we keep it in `type session`? We can just use the `Config.get Auth` whenever needed.

File: src/engine.ml (hunk 3 above)
Comment (reynir) [reply src/engine.ml:111]: Yea, that makes sense. That's what openvpn does too in fact!

**Comment by hannesm** [comment]
fixes #82

File: src/state.ml (hunk 4 above)
Comment (reynir) [reply src/state.ml:196]: I find the story in openvpn about cipher (and hmac?) negotiation hard to follow. At the moment I don't remember any negotiation about `auth` - except if the cipher is AEAD where `auth` is ignored.

I added it to `session` because we seem to push that "further" than `config`. I also had in mind that we need to choose SHA256 as `auth` for control packets when using tls-crypt-v2, but I'm not sure how I got this idea and I don't think it is the case.

File: src/config.ml (hunk 1 above)
Comment (reynir) [reply src/config.ml:250]: I have not tried that yet. With main branch I get errors on the openvpn 2.6 server such as `TLS Error: unknown opcode received from [AF_INET]192.168.1.84:58614 op=12` and a few of `TLS Error: client->client or server->server connection attempted from [AF_INET]192.168.1.84:58614`. It seems to get stuck after the handshake. I should test this because if I can get as far with a different `auth` we exercise the code path and it's no worse.

Comment (hannesm) [reply src/config.ml:250]: oh, even after #98 you're not able to establish a connection?

Comment (reynir) [reply src/config.ml]: With the other issues fixed I tried with `auth sha256`, and that does not seem to work: `TLS Error: cannot locate HMAC in incoming packet from [AF_INET]192.168.1.84:37694`

File: src/state.ml (hunk 4 above)
Comment (reynir) [reply src/state.ml]: I have refactored the code to not store the hmac algorithm in `type session`.

I have not seen anything that suggests the hmac algorithm can be negotiated - other than if you use an AEAD cipher. We probably need to restructure the code in order to support AEAD. I will not think about that right now though.

**Comment by reynir** [comment]
Now rebased on main and ready for review / merge.

File: src/config.ml (hunk 1 above)
Comment (reynir) [reply src/config.ml]: It was due to an optional argument with default value `` `SHA1`` that was never passed.
